### PR TITLE
【Hackathon 10th Spring No.1】Add `dilation` option for MaxPool1D/2D/3D -part

### PR DESCRIPTION
**Pull request by jinyouzhi**
### PR Category

User Experience

### PR Types

New features

### Description

 **XPU 暂时不修改**  （需xpu sdk 侧支持）

### 是否引起精度变化

否

### CODE REVIEW

**Comment by jinyouzhi** [comment]
ready to review

**Comment by jinyouzhi** [comment]
@zhwesky2010 研发大佬可以帮忙review一下吗？converage test达标了

File: python/paddle/nn/functional/pooling.py [hunk 1]
```diff
@@ -643,10 +648,21 @@ def max_pool1d(
     # use 2d to implement 1d should expand padding in advance.
     padding = _expand_low_nd_padding(padding)
 
+    # Check if dilation is non-trivial (not all ones)
+    has_dilation = any(d != 1 for d in dilation_list)
+
     if in_dynamic_or_pir_mode():
-        if return_mask:
+        if return_mask or has_dilation:
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/pooling.py:658]: 这个还需要判断has_dilation吗。这个默认是1，默认情况会切换到下面的_C_ops.pool2d分支里去，会导致不兼容吧

File: python/paddle/nn/functional/pooling.py [hunk 2]
```diff
@@ -1229,10 +1257,30 @@ def max_pool2d(
             "When setting return_mask to true, data_format must be set to NCHW in API:max_pool2d"
         )
 
+    # Check if dilation is non-trivial (not all ones)
+    dilation = (
+        dilation if isinstance(dilation, (list, tuple)) else [dilation] * 2
+    )
+    has_dilation = any(d != 1 for d in dilation)
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/pooling.py:1270]: 同上

File: python/paddle/nn/functional/pooling.py [hunk 3]
```diff
@@ -1386,10 +1444,30 @@ def max_pool3d(
             "When setting return_mask to true, data_format must be set to NCDHW in API:max_pool3d"
         )
 
+    # Check if dilation is non-trivial (not all ones)
+    dilation = (
+        dilation if isinstance(dilation, (list, tuple)) else [dilation] * 3
+    )
+    has_dilation = any(d != 1 for d in dilation)
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/pooling.py:1460]: 同上

File: python/paddle/nn/functional/pooling.py (hunk 1 above)
Comment (jinyouzhi) [reply python/paddle/nn/functional/pooling.py:658]: > 这个还需要判断has_dilation吗。

根据652行，has_dilation是当dilation != 1 时取true，支持dilation逻辑写在max_pool2d_with_index这里面

> 这个默认是1，默认情况会切换到下面的_C_ops.pool2d分支里去，会导致不兼容吧

是的，只有默认的dilation=1会走pool2d并且忽略dilation这个参数。

**Comment by jinyouzhi** [comment]
@zhwesky2010 请教一下，像这个 XPU 的 kernel实现 在哪可以找到？
https://github.com/PaddlePaddle/Paddle/blob/14867a1059226160125cf772ed85984140d87b3a/paddle/phi/kernels/xpu/pool_kernel.cc#L368
进一步开发是不是要参考paddle/phi/kernels/xpu/plugin 这里的机制来开发XPU kernel？

File: python/paddle/nn/functional/pooling.py (hunk 3 above)
Comment (zhwesky2010) [reply python/paddle/nn/functional/pooling.py:1460]: @jinyouzhi 默认分支最好不要发生变化，不用判断has_dilation，默认的1也直接传入到dilation里就行？

**Comment by zhwesky2010** [comment]
> @zhwesky2010 请教一下，像这个 XPU 的 kernel实现 在哪可以找到？
> 
> https://github.com/PaddlePaddle/Paddle/blob/14867a1059226160125cf772ed85984140d87b3a/paddle/phi/kernels/xpu/pool_kernel.cc#L368
> 
> 
> 进一步开发是不是要参考paddle/phi/kernels/xpu/plugin 这里的机制来开发XPU kernel？

这个是xpu的sdk支持的，如果底层sdk不支持，可以先按抛出异常来处理

File: python/paddle/nn/functional/pooling.py (hunk 3 above)
Comment (jinyouzhi) [reply python/paddle/nn/functional/pooling.py:1460]: 哦哦，是这样，只改了max_pool2d_with_index是因为它是专属于max_pool2d的kernel，其他情况落到通用的pool2d。

这里根据has_dilation分发是必要的，去掉这个判断需要把else这边的pool2d也支持上dilation。pool2d是一个通用的max/min/avg的，对应好几个算子，改动面有点大 @zhwesky2010

**Comment by jinyouzhi** [comment]
> > @zhwesky2010 请教一下，像这个 XPU 的 kernel实现 在哪可以找到？
> 
> > 
> 
> > https://github.com/PaddlePaddle/Paddle/blob/14867a1059226160125cf772ed85984140d87b3a/paddle/phi/kernels/xpu/pool_kernel.cc#L368
> 
> > 
> 
> > 
> 
> > 进一步开发是不是要参考paddle/phi/kernels/xpu/plugin 这里的机制来开发XPU kernel？
> 
> 
> 
> 这个是xpu的sdk支持的，如果底层sdk不支持，可以先按抛出异常来处理

明白了！不好意思，我没搜到这个xpu sdk的文档，在哪里能确认底层支持情况？目前是抛出异常的

File: paddle/phi/kernels/funcs/pooling.cc [hunk 4]
```diff
@@ -1146,6 +1162,11 @@ class MaxPool2dWithIndexFunctor<CPUContext, T1, T2> {
           if (adaptive) {
             hstart = AdaptStartIndex(ph, input_height, output_height);
             hend = AdaptEndIndex(ph, input_height, output_height);
+          } else if (dilation) {
```
Comment (zhwesky2010) [inline paddle/phi/kernels/funcs/pooling.cc]: 这里是不是这样判断更好：
```
else if (dilation_height != 1):
    ...
```

如果遇到 dilation = [1, 2] 这种，所有分支都需要切换到dilation分支吗

File: test/legacy_test/test_max_pool_dilation.py [hunk 5]
```diff
@@ -0,0 +1,2483 @@
...
+    """Test MaxPool1D with dilation parameter on CPU."""
+
+    def setUp(self):
+        np.random.seed(123)
+        self.places = [paddle.CPUPlace()]
+
+    def test_max_pool1d_dilation_functional(self):
+        """Test F.max_pool1d with dilation parameter."""
+        for place in self.places:
+            paddle.set_device('cpu')
```
Comment (zhwesky2010) [inline test/legacy_test/test_max_pool_dilation.py]: 测试这里不需要单独设置某个device，有几十条流水线，包括各种device上都有

File: test/legacy_test/test_max_pool_dilation.py [hunk 6]
```diff
@@ -0,0 +1,2483 @@
...
+
+        np.testing.assert_allclose(
+            result.numpy(),
+            expected,
+            rtol=1e-05,
+            err_msg="MaxPool1D layer dilation with ceil_mode mismatch",
+        )
+
+
+class TestMaxPool2DLayerDilation(unittest.TestCase):
```
Comment (zhwesky2010) [inline test/legacy_test/test_max_pool_dilation.py:1000]: 提交一下paconvert的本地测试截图，增强下paconvert中maxpool1d/2d/3d的测试case，确保和pytorch运算结果一致。

File: test/legacy_test/test_pool_max_op.py [hunk 7]
```diff
@@ -147,8 +147,16 @@ def max_pool3d_with_index_wrapper(
     adaptive=False,
```
Comment (zhwesky2010) [inline test/legacy_test/test_pool_max_op.py:147]: 测试了OP反向没有，怎么确保反向计算结果是正确的，可以加个max_pool2d_with_index / max_pool3d_with_index的OPTeST，通过test_check_grad会自动测试反向

infoflow 2026-01-29 16-48-53

File: paddle/phi/kernels/funcs/pooling.cc (hunk 4 above)
Comment (jinyouzhi) [reply paddle/phi/kernels/funcs/pooling.cc]: great idea, done.

File: test/legacy_test/test_max_pool_dilation.py (hunk 5 above)
Comment (jinyouzhi) [reply test/legacy_test/test_max_pool_dilation.py]: 好的，重构了下，合并了cpu/gpu UT，删掉了静态图的ut（好像完全废弃了？完全跑不到）

File: test/legacy_test/test_max_pool_dilation.py (hunk 6 above)
Comment (jinyouzhi) [reply test/legacy_test/test_max_pool_dilation.py:1000]: image

运算结果是一致的，mask dtype 不一致，后续在另一个PR修复：

* https://github.com/PaddlePaddle/Paddle/pull/77529

File: test/legacy_test/test_pool_max_op.py (hunk 7 above)
Comment (jinyouzhi) [reply test/legacy_test/test_pool_max_op.py:147]: 已加。

**Comment by YqGe585** [comment]
IXUCA rerun仍然失败的,辛苦rebase到develop后重新触发CI

File: test/legacy_test/test_max_pool_dilation.py [hunk 8]
```diff
@@ -0,0 +1,1647 @@
...
+            return_mask=True,
+        )
+
+        # Backward
+        loss = paddle.mean(result)
+        loss.backward()
+
+        # Check gradient exists and has correct shape
+        self.assertIsNotNone(input_tensor.grad)
+        self.assertEqual(list(input_tensor.grad.shape), [2, 3, 16, 16])
```
Comment (zhwesky2010) [inline test/legacy_test/test_max_pool_dilation.py:812]: 这个需要测一下反向的值，看一下能否通过OpTest来测，test_check_grad会自动求expect_grad。

Comment (jinyouzhi) [reply test/legacy_test/test_max_pool_dilation.py:812]: OK, OpTest的算子添加在后面了，test_check_grad 加在 1445/1599 行

**Comment by jinyouzhi** [comment]
> IXUCA rerun仍然失败的,辛苦rebase到develop后重新触发CI

好的，同一流水线有其他的在跑没re-run起来，anyway已经rebase了

**Comment by jinyouzhi** [comment]
除了static和Approval 只剩Linux-DCU这个CI挂了，看起来挂在dcc编译pooling.cu，加了个commit把while循环尝试修复
image

**Comment by zhwesky2010** [comment]
@jinyouzhi DCU问题修复了吗

**Comment by jinyouzhi** [comment]
> @jinyouzhi DCU问题修复了吗

要等ci结果，~~dcu的ci因为网络fail了，能帮我re-run吗？我这边应该只能等流水线内所有都跑完才能rerun~~（已rerun了

**Comment by jinyouzhi** [comment]
> @jinyouzhi DCU问题修复了吗

好了编过了，居然build花了2.5小时

**Comment by jinyouzhi** [comment]
@zhwesky2010 DCU CI pass了

**Comment by jinyouzhi** [comment]
@zhwesky2010 CI都过了，麻烦您再review一下

File: test/legacy_test/test_max_pool_dilation.py [hunk 9]
```diff
@@ -0,0 +1,1647 @@
...
+        self.shape = [2, 3, 16, 16]
+        self.ksize = [3, 3]
+        self.strides = [2, 2]
+        self.paddings = [1, 1]
+        self.dilations = [2, 2]
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_check_grad(self):
```
Comment (jinyouzhi) [inline test/legacy_test/test_max_pool_dilation.py:1404]: @zhwesky2010 基于OpTest的check_grad加上了

File: paddle/phi/kernels/funcs/pooling.cu [hunk 10]
```diff
@@ -2323,8 +2379,8 @@ class MaxPool2dWithIndexFunctor<phi::GPUContext, T1, T2> {
 
 /*
  * All tensors are in NCHW format.
- * Ksize, strides, paddings are two elements. These two elements represent
- * height and width, respectively.
+ * Ksize, strides, paddings, dilations are two elements. These two elements
+ * represent height, width and dilation, respectively.
```
Comment (zhwesky2010) [inline paddle/phi/kernels/funcs/pooling.cu]: 这个注释不太对，还是
`These two elements represent height, width`

File: paddle/phi/kernels/funcs/pooling.cu [hunk 11]
```diff
@@ -2564,8 +2665,8 @@ __global__ void KernelMaxPool3DWithIdxGrad(
 
 /*
  * All tensors are in NCDHW format.
- * Ksize, strides, paddings are three elements. These three elements represent
- * depth, height and width, respectively.
+ * Ksize, strides, paddings, dilations are three elements. These three elements
+ * represent depth, height, width and dilation, respectively.
```
Comment (zhwesky2010) [inline paddle/phi/kernels/funcs/pooling.cu]: 这个注释不太对，还是
`These two elements represent height, width and dilation`

**Comment by jinyouzhi** [comment]
@zhwesky2010 可以再review一下吗？DCU问题修复了

File: test/legacy_test/test_max_pool_dilation.py [hunk 12]
```diff
@@ -0,0 +1,1647 @@
...
+
+    def init_test_case(self):
+        self.shape = [1, 2, 12, 14]
+        self.ksize = [2, 2]
+        self.strides = [2, 2]
+        self.paddings = [0, 0]
+        self.dilations = [2, 3]
+
+
+@unittest.skipIf(
```
Comment (zhwesky2010) [inline test/legacy_test/test_max_pool_dilation.py]: 如果整个单测都需要@unittest.skipIf(，考虑从XPU/CMakeLists.txt进行控制。避免整个文件的大量skipif

**Comment by zhwesky2010** [comment]
@jinyouzhi 小问题，尽快修改后合入。

**Comment by jinyouzhi** [comment]
> @jinyouzhi 小问题，尽快修改后合入。

好的改完了， XPU 跳过 ut 改成在legacy_test/CMakeLists.txt加这个了
```
--- a/test/legacy_test/CMakeLists.txt
+++ b/test/legacy_test/CMakeLists.txt
@@ -91,6 +91,10 @@ if(NOT WITH_PYTHON AND ON_INFER)
   list(REMOVE_ITEM TEST_OPS test_eager_trace_op)
 endif()

+if(WITH_XPU)
+  list(REMOVE_ITEM TEST_OPS test_max_pool_dilation)
+endif()
+
 list(REMOVE_ITEM TEST_OPS test_fused_attention_pass)
 list(REMOVE_ITEM TEST_OPS test_fused_feedforward_pass)
 if(NOT WITH_GPU)
```
https://github.com/PaddlePaddle/Paddle/pull/77495/commits/f165b04f9de7ac517432c31f56627942c062629f

**Review by zhwesky2010 — DISMISSED** [review]
LGTM

File: python/paddle/nn/layer/pooling.py [hunk 13]
```diff
@@ -663,11 +665,16 @@ class MaxPool1D(Layer):
             >>> print(indices.shape)
             paddle.Size([1, 3, 16])
 
+            >>> MaxPool1D = nn.MaxPool1D(kernel_size=2, stride=2, padding=0, dilation=2)
```
Comment (SigureMo) [inline python/paddle/nn/layer/pooling.py]:
```suggestion
            >>> max_pool1d = nn.MaxPool1D(kernel_size=2, stride=2, padding=0, dilation=2)
```

实例用 `snake_case`，上面一起改一下吧，其他 max pool 示例代码也一样

Comment (jinyouzhi) [reply python/paddle/nn/layer/pooling.py]: ok，已修改。

**Review by zhwesky2010 — APPROVED** [review]

**Review by SigureMo — APPROVED** [review]

File: python/paddle/nn/layer/pooling.py [hunk 14]
```diff
@@ -97,8 +97,8 @@ class AvgPool1D(Layer):
             >>> import paddle.nn as nn
 
             >>> data = paddle.uniform([1, 3, 32], dtype="float32", min=-1, max=1)
-            >>> AvgPool1D = nn.AvgPool1D(kernel_size=2, stride=2, padding=0)
-            >>> pool_out = AvgPool1D(data)
+            >>> avg_pool1d = nn.AvgPool1D(kernel_size=2, stride=2, padding=0)
+            >>> pool_out = avg_pool1D(data)
```
Comment (algorithm1832) [inline python/paddle/nn/layer/pooling.py:101]: @jinyouzhi 

R101的`avg_pool1D`应为`avg_pool1d`喵

Comment (jinyouzhi) [reply python/paddle/nn/layer/pooling.py:101]: 谢谢提醒！@algorithm1832 
https://github.com/PaddlePaddle/Paddle/pull/77738

**Comment by zhwesky2010** [comment]
@jinyouzhi 这个修改经过内部测试会有不兼容影响。原因是dilation插入到return_mask之前，导致原来的return_mask作为位置参数时无法使用。因此需要改成采用 **使用装饰器支持签名重载** 的API升级方案，修改方式为：

先将dilation放到paddle参数表末尾，确保paddle自身兼容，签名如下：

`paddle.nn.functional.max_pool2d(x, kernel_size, stride, padding, return_mask, celi_mode, dilation, data_format)`

`torch.nn.functional.max_pool2d(input, kernel_size, stride, padding, dilation, celi_mode, return_indices)`

然后通过装饰器，实现torch用法对齐：处理 **两个参数名不一致+后面三个参数顺序不一致** 的问题。新增一个`maxpool_decorator`，判断第5个参数是否是bool，如果不是bool，则按torch签名来匹配第5/6/7个位置参数，如果是bool则不做处理。具体实现逻辑可以参考index_add_decotator：
https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/utils/decorator_utils.py#L900

最后在PaConvert上验证可以与Pytorch用法一致，提交PR的本地测试截图。待Paddle合入后次日合入PaConveert。

**Comment by jinyouzhi** [comment]
> @jinyouzhi 这个修改经过内部测试会有不兼容影响。原因是dilation插入到return_mask之前，导致原来的return_mask位置参数用法无法使用。因此需要改成采用 **使用装饰器支持签名重载** 的API升级方案，修改方式为：
> 
> 
> 
> 先将dilation放到paddle参数表末尾，确保paddle自身兼容，签名如下：
> 
> 
> 
> `paddle.nn.functional.max_pool2d(x, kernel_size, stride, padding, return_mask, celi_mode, dilation, data_format)`
> 
> 
> 
> `torch.nn.functional.max_pool2d(input, kernel_size, stride, padding, dilation, celi_mode, return_indices)`
> 
> 
> 
> 然后通过装饰器，实现torch用法对齐：处理 **两个参数名不一致+后面三个参数顺序不一致** 的问题。新增一个`maxpool_decorator`，判断第5个参数是否是bool，如果不是bool，则按torch签名来匹配第5/6/7个位置参数，如果是bool则不做处理。具体实现逻辑可以参考index_add_decotator：
> 
> https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/utils/decorator_utils.py#L900
> 
> 
> 
> 最后在PaConvert上验证可以与Pytorch用法一致，提交PR的本地测试截图。待Paddle合入后次日合入PaConveert。

收到
